### PR TITLE
Remove node access from virt-handler

### PR DESCRIPTION
**Pull request by xpivarc**
**What this PR does / why we need it**:
As of today a `virt-handler` has "full" access to all `Node`s. Therefore installing Kubevirt means that every node will have a service account that can be miss used to influence the scheduling of other Pods, potentially with a more privileged service account.
This PR introduces `ShadowNode` CRD, intended for internal use for `virt-handler`. The `virt-handler` will keep read access to nodes but will lose write access. Write will be "proxied" through `ShadowNode` object and write to the actual `Node`  will be done by virt-controller who will filter what type of labels and annotations will be proxied.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
See https://github.com/kubevirt/community/pull/246

There are a few cons to the approach. Creating 1:1 shadow nodes to nodes and therefore increasing the #api calls we do and storage we take. 
Because we "proxy" the node updates we now depend on the availability of virt-controller and we increase the work required by virt-controller. In the worst case, we can see heartbeat timeout more in corner cases. The latency of sync increases because we remove the "test" on the node.

For less critical issues, right now we remove any `.* kubevirt.io .*` label and annotation. This might break a few cases where we or others use these labels to enhance scheduling or other functionality. 

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by rthallisey** [comment]
>This PR introduces ShadowNode CRD, intended for internal use for virt-handler. The virt-handler will keep read access to nodes but will lose write access. Write will be "proxied" through ShadowNode object and write to the actual Node will be done by virt-controller who will filter what type of labels and annotations will be proxied.
>
This idea sounds like you're putting a shim between the RBAC policy and RBAC controller, similar to the way the scheduler is broken down to pluggable steps.   This might be a useful KEP.

**Comment by rmohr** [comment]
> There are a few cons to the approach. Creating 1:1 shadow nodes to nodes and therefore increasing the #api calls we do and storage we take. Because we "proxy" the node updates we now depend on the availability of virt-controller and we increase the work required by virt-controller. In the worst case, we can see heartbeat timeout more in corner cases. 

I noticed that virt-handler is now pushing the heartbeat to the shadow node, but the controller side still seems to take the heartbeat timestamp on the node as the source of truth and there is really just a "mirror" injected?

If that is not what is happening already, I would suggest the following:

 1. there should be no reason anymore for us to put annotations on the node. Everything which is now an annotation on the node should only be used internally by virt-controller, and the virt-controller (especially node.go) can now directly read the heartbeat from the shadow node. As such there is not increase of likelyhood of seeing heartbeat issues. The shadow node also has a proper status section now. A heartbeat annotation would be a great fit for a status condition. Finally we can shape the API :)
 2. there shold only be a need to synchronize labels for scheduling purposes to the real node
 3. Since we can't start VMs if virt-controller is down, there should not be really a con left. If virt-controller is down, true nothing gets synced, but also nothing can start anyway.
 
> The latency of sync increases because we remove the "test" on the node.

I may figure out what you mean when finally having checked the whole PR, but could you outline what you mean by that?


> 
> For less critical issues, right now we remove any `.* kubevirt.io .*` label and annotation. This might break a few cases where we or others use these labels to enhance scheduling or other functionality.

`*.kubevirt.io/` is our domain. Synchronizing on it should be fine. I can also see that some people may have chosen to use our label domain as well, but I think owning such a label domain is a well enough established convention to delete unknown labels.

**Comment by rmohr** [comment]
I am btw. also very keen on not having to continue to periodically update an annotation on the node. I have over the years quite a few issues around that. Not that we do anythying wrong or uncommon here, actually always other controllers were the root cause (like too long reconcile periods, just overriding annotations, ...) but less surface which can cause issues is great.

**Comment by xpivarc** [comment]
> This idea sounds like you're putting a shim between the RBAC policy and RBAC controller, similar to the way the scheduler is broken down to pluggable steps. This might be a useful KEP.

@rthallisey I will update the PR description with some alternatives. Generally speaking, a Kubernetes solution for this problem could be delivered in a year or two(Note this problem is not specific to Kubevirt). Therefore I think this is a viable short-term improvement/fix.

**Comment by xpivarc** [comment]
@rmohr PTAL, I noticed we do miss rolback e2e test. Please double-check check this has a chance to work.

**Comment by xpivarc** [comment]
Note for me: I need to update the description outlining the approach.

**Comment by rmohr** [comment]
Out for a week. Will review next week.

**Comment by xpivarc** [comment]
/hold
Seeing DaemonSet killing all DSes. Somehow I triggered some edge cases.

**Comment by maiqueb** [comment]
@xpivarc o/

Is there a design doc or something similar outlining the solution ?

**Comment by xpivarc** [comment]
> @xpivarc o/
> 
> Is there a design doc or something similar outlining the solution ?

There is nothing online available. Do you have any concerns? Would you like to see one?

**Comment by maiqueb** [comment]
> > @xpivarc o/
> > Is there a design doc or something similar outlining the solution ?
> 
> There is nothing online available. Do you have any concerns? Would you like to see one?

I don't have any concerns at the moment, but I spent 0 seconds thinking about this problem.

I would like to see one. Isn't that the common practice for a new feature ? You're even adding a new CRD IIUC.

**Comment by xpivarc** [comment]
> > > @xpivarc o/
> > > Is there a design doc or something similar outlining the solution ?
> > 
> > 
> > There is nothing online available. Do you have any concerns? Would you like to see one?
> 
> I don't have any concerns at the moment, but I spent 0 seconds thinking about this problem.
> 
> I would like to see one. Isn't that the common practice for a new feature ? You're even adding a new CRD IIUC.

Not necessary, we usually do a design document/proposal only for bigger features. This is more of a "bug" fix, an internal change that does not have user-facing changes. I can draft something up if there is a need for it.

**Comment by xpivarc** [comment]
Flake is not related - see https://github.com/kubevirt/kubevirt/pull/10417

File: pkg/virt-handler/heartbeat/heartbeat.go [hunk 1]
```diff
@@ -24,25 +27,30 @@ import (
 const failedSetCPUManagerLabelFmt = "failed to set a cpu manager label on host %s"
 
 type HeartBeat struct {
-	clientset                 k8scli.CoreV1Interface
+	clientSet                 k8scli.NodeInterface
+	client                    kubecli.ShadowNodeInterface
```
Comment (lzang) [inline pkg/virt-handler/heartbeat/heartbeat.go:31]: Please make the variable name to be more distinguishable, like nodeClient and shadowNodeClient

File: pkg/virt-handler/heartbeat/heartbeat.go [hunk 2]
```diff
@@ -93,7 +101,13 @@ func (h *HeartBeat) labelNodeUnschedulable() (done chan struct{}) {
 			v1.CPUManager, cpuManagerEnabled,
 			v1.VirtHandlerHeartbeat, string(now),
 		))
-		_, err = h.clientset.Nodes().Patch(context.Background(), h.host, types.StrategicMergePatchType, data, metav1.PatchOptions{})
+
+		// TODO: This can be safely removed once we do not support upgrade from a version that does not have Shadow Node
+		_, err = h.clientSet.Patch(context.Background(), h.host, types.StrategicMergePatchType, data, metav1.PatchOptions{})
```
Comment (lzang) [inline pkg/virt-handler/heartbeat/heartbeat.go:106]: Should this be controlled by nodeBackoff flag as well?

File: pkg/virt-handler/heartbeat/heartbeat.go [hunk 3]
```diff
@@ -24,25 +27,30 @@ import (
...
+		clientSet:               nodeClient,
 		deviceManagerController: deviceManager,
 		clusterConfig:           clusterConfig,
 		host:                    host,
 		// This is a temporary workaround until k8s bug #66525 is resolved
 		cpuManagerPaths:           []string{virtutil.CPUManagerPath, virtutil.CPUManagerOS3Path},
 		devicePluginPollIntervall: 1 * time.Second,
 		devicePluginWaitTimeout:   10 * time.Second,
+
+		nodeBackoff: true,
```
Comment (lzang) [inline pkg/virt-handler/heartbeat/heartbeat.go:53]: Add a comment on what this flag means or change it to a more intuitive name, such as updateNodeHeartBeat

File: pkg/virt-handler/node-labeller/node_labeller.go [hunk 4]
```diff
@@ -231,31 +228,63 @@ func skipNodeLabelling(node *v1.Node) bool {
...
-			Path:  "/metadata/labels",
-			Value: node.Labels,
-		})
+		}, replaceLabels(node.Labels))
+
+		if len(node.Annotations) == 0 {
+			node.Annotations = map[string]string{}
+		}
+		node.Annotations["kubevirt.io/shadownode"] = string(now)
+		shadownodeOperation = append(shadownodeOperation, replaceLabels(node.Labels))
```
Comment (lzang) [inline pkg/virt-handler/node-labeller/node_labeller.go:267]: Why here you don't need originalNode.Labels like in line 237 for node? Is that because you assume you can do a replace here without check?

File: cmd/virt-handler/virt-handler.go [hunk 5]
```diff
@@ -229,6 +238,15 @@ func (app *virtHandlerApp) Run() {
 		panic(err)
 	}
 
+	_, err = app.virtCli.ShadowNodeClient().Create(context.TODO(), &v1.ShadowNode{
+		ObjectMeta: metav1.ObjectMeta{
```
Comment (lzang) [inline cmd/virt-handler/virt-handler.go:242]: If you create this object with virt-handler, who will take care of the deletion of this object in case of a node removal?

File: pkg/virt-operator/resource/generate/rbac/handler.go [hunk 6]
```diff
@@ -89,7 +100,6 @@ func newHandlerClusterRole() *rbacv1.ClusterRole {
 					"nodes",
 				},
 				Verbs: []string{
-					"patch",
```
Comment (lzang) [inline pkg/virt-operator/resource/generate/rbac/handler.go:92]: Remove the permission means the node patching will fail.

File: hack/generate.sh [hunk 7]
```diff
@@ -50,7 +50,7 @@ fi
 
 client-gen --clientset-name versioned \
     --input-base kubevirt.io/api \
-    --input export/v1alpha1,snapshot/v1alpha1,instancetype/v1alpha1,instancetype/v1alpha2,instancetype/v1beta1,pool/v1alpha1,migrations/v1alpha1,clone/v1alpha1 \
+    --input core/v1,export/v1alpha1,snapshot/v1alpha1,instancetype/v1alpha1,instancetype/v1alpha2,instancetype/v1beta1,pool/v1alpha1,migrations/v1alpha1,clone/v1alpha1 \
```
Comment (lzang) [inline hack/generate.sh:53]: Why mixing this v1 change with the shadow node changes? It is worth to be a separate CL in my opinion.

File: pkg/virt-controller/watch/node.go [hunk 8]
```diff
@@ -33,28 +33,28 @@ const (
...
-	clientset        kubecli.KubevirtClient
-	Queue            workqueue.RateLimitingInterface
-	nodeInformer     cache.SharedIndexInformer
-	vmiInformer      cache.SharedIndexInformer
-	recorder         record.EventRecorder
-	heartBeatTimeout time.Duration
-	recheckInterval  time.Duration
+	clientset          kubecli.KubevirtClient
+	Queue              workqueue.RateLimitingInterface
+	shadowNodeInformer cache.SharedIndexInformer
```
Comment (lzang) [inline pkg/virt-controller/watch/node.go:38]: So both the node controller and shadow node controller are watching shadow node changes?  Does it worth to merge them?

File: pkg/virt-handler/node-labeller/node_labeller.go [hunk 9]
```diff
@@ -231,31 +228,63 @@ func skipNodeLabelling(node *v1.Node) bool {
...
-			Op:    "replace",
-			Path:  "/metadata/labels",
-			Value: node.Labels,
-		})
+		}, replaceLabels(node.Labels))
+
+		if len(node.Annotations) == 0 {
+			node.Annotations = map[string]string{}
+		}
+		node.Annotations["kubevirt.io/shadownode"] = string(now)
```
Comment (lzang) [inline pkg/virt-handler/node-labeller/node_labeller.go:266]: What is the usage of this annotation?

**Comment by xpivarc** [comment]
@maiqueb @lzang I have created a design doc https://github.com/kubevirt/community/pull/246

File: pkg/virt-controller/watch/shadow_node.go [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+	nodeAnnotations := node.Annotations
+
+	shadowNodeAnnotations := filterKubevirt(shadowNode.Annotations)
+
+	filteredNodeAnnotations := filterNotKubevirt(nodeAnnotations)
+
+	for key, value := range shadowNodeAnnotations {
+		filteredNodeAnnotations[key] = value
+	}
+
```
[the remark this answers is not in the document]
Comment (xpivarc) [reply pkg/virt-controller/watch/shadow_node.go:131]: Will do

File: pkg/virt-handler/heartbeat/heartbeat.go (hunk 1 above)
Comment (xpivarc) [reply pkg/virt-handler/heartbeat/heartbeat.go:31]: Agree

File: cmd/virt-handler/virt-handler.go (hunk 5 above)
Comment (xpivarc) [reply cmd/virt-handler/virt-handler.go:242]: Thanks for pointing this out. We can either use the Kube garbage collector or I can handle this in the shadow node controller.  I don't see any benefit of one over the other, wdyt?

File: hack/generate.sh (hunk 7 above)
Comment (xpivarc) [reply hack/generate.sh:53]: I am using the generated client. What do you mean by CL?

File: pkg/virt-controller/watch/node.go (hunk 8 above)
Comment (xpivarc) [reply pkg/virt-controller/watch/node.go:38]: I would leave this out of the PR but it is likely we could merge the logic of these 2.

File: pkg/virt-handler/heartbeat/heartbeat.go (hunk 3 above)
Comment (xpivarc) [reply pkg/virt-handler/heartbeat/heartbeat.go:53]: Agree

File: pkg/virt-handler/heartbeat/heartbeat.go (hunk 2 above)
Comment (xpivarc) [reply pkg/virt-handler/heartbeat/heartbeat.go:106]: `nodeBackoff` is meant only to decrease #writes. This function is called rarely and therefore I don't think we need to include it.

File: pkg/virt-operator/resource/generate/rbac/handler.go (hunk 6 above)
Comment (xpivarc) [reply pkg/virt-operator/resource/generate/rbac/handler.go:92]: Yes, this is intentional.

**Comment by xpivarc** [comment]
@lzang I hope I answered all of your comments. I will rebase this PR in the upcoming days.

File: cmd/virt-handler/virt-handler.go [hunk 11]
```diff
@@ -181,9 +182,17 @@ func (app *virtHandlerApp) prepareCertManager() (err error) {
 	return
 }
 
+// NODE labels
 func (app *virtHandlerApp) markNodeAsUnschedulable(logger *log.FilteredLogger) {
 	data := []byte(fmt.Sprintf(`{"metadata": { "labels": {"%s": "false"}}}`, v1.NodeSchedulable))
+
+	// TODO: This can be safely removed once we do not support upgrade from a version that does not have Shadow Node
```
[the remark this answers is not in the document]
Comment (lzang) [reply cmd/virt-handler/virt-handler.go:189]: > We are not going to modify the upgrade path with virt-operator. As of now, new CRDs are created, a handler is updated first, followed by controller, and in the end, the RBAC will be removed automatically.
> 
> Therefore the writes we leave in the handler should be failing only after the upgrade is successful.

But for the handler to watch/patch the new CRD, you need the RBAC to be updated first, right?

File: cmd/virt-handler/virt-handler.go (hunk 5 above)
Comment (lzang) [reply cmd/virt-handler/virt-handler.go:242]: By using kube garbage collector, do you mean let the node owns the object?
Shadow-node controller sounds slightly better to me as we can follow the general principle of whoever creates the object will take care of the clean up.

Another question is if the shadow node already exists, do we reconcile its labels to make sure they are up-to-date later?

File: hack/generate.sh (hunk 7 above)
Comment (lzang) [reply hack/generate.sh:53]: CL means PR in this context.

File: cmd/virt-handler/virt-handler.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (xpivarc) [reply cmd/virt-handler/virt-handler.go:189]: Yes, we do create the new RBAC at first and keep the old around while the upgrade is happening.

File: cmd/virt-handler/virt-handler.go (hunk 5 above)
Comment (xpivarc) [reply cmd/virt-handler/virt-handler.go:242]: >By using kube garbage collector, do you mean let the node owns the object? 

I think the other way around would be better.

>Another question is if the shadow node already exists, do we reconcile its labels to make sure they are up-to-date later?

I don't follow. Can you elaborate?

File: pkg/virt-handler/node-labeller/node_labeller.go (hunk 4 above)
Comment (xpivarc) [reply pkg/virt-handler/node-labeller/node_labeller.go:267]: Good question, I think this is wrong as shadownode could have more information than node and I could replace it/lose it.
I think I need to rework this in a way that I individually construct `add`/`replace`/`delete` patches.

**Comment by xpivarc** [comment]
@lzang PTAL

File: cmd/virt-handler/virt-handler.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (lzang) [reply cmd/virt-handler/virt-handler.go:190]: But aren't they have the same name? Curious how can you have both the new and the old RBAC together.

File: cmd/virt-handler/virt-handler.go (hunk 5 above)
Comment (lzang) [reply cmd/virt-handler/virt-handler.go:243]: For example, when you have an upgrade case where some labels are obsoleted, since you already have the old shadow-node there, so you will not recreate it, and then the old shadow-node may have the obsoleted labels.  Are you going to reconcile at some time and remove those obsoleted labels?

**Comment by lzang** [comment]
@xpivarc overall looks good to me as long as the upgrade case is tested, i.e there is a transitional period where both nodes and shadownodes are updated before the virt-controller is updated.

**Comment by xpivarc** [comment]
> @xpivarc overall looks good to me as long as the upgrade case is tested, i.e there is a transitional period where both nodes and shadownodes are updated before the virt-controller is updated.

The upgrade is covered by the existing test.

**Comment by alaypatel07** [comment]
@xpivarc this PR came up for discussion in couple of sigs (sig-scale and sig-api). Because this is a somewhat larger change and affects multiple CRDs (nodes, shallownodes), I wonder if we need a design doc explaining the advantage from the change, seemless upgrades and scale impact from the change

/cc @EdDev @mooperd

**Comment by xpivarc** [comment]
> @xpivarc this PR came up for discussion in couple of sigs (sig-scale and sig-api). Because this is a somewhat larger change and affects multiple CRDs (nodes, shallownodes), I wonder if we need a design doc explaining the advantage from the change, seemless upgrades and scale impact from the change
> 
> /cc @EdDev @mooperd

I have updated the description.

**Comment by mooperd** [comment]
@xpivarc would you be interested in running through this PR with me in a call? I'd really like to understand the attack surface more and what exactly these changes are intended to mitigate.

**Comment by xpivarc** [comment]
> @xpivarc would you be interested in running through this PR with me in a call? I'd really like to understand the attack surface more and what exactly these changes are intended to mitigate.

This PR is targeting https://github.com/kubevirt/kubevirt/security/advisories/GHSA-cp96-jpmq-xrr2. If you have any questions you can find me on Kubernetes Slack under [@Luboslav Pivarc](https://kubernetes.slack.com/team/UTYB1JSSC)

**Comment by mooperd** [comment]
Thanks @xpivarc. I am a bit confused about this line:

`If a malicious user has taken over a Kubernetes node where virt-handler (the KubeVirt node-daemon) is running, the virt-handler service account can be used to modify all node specs.`

Please could you clarify it? If a node is compromised then all virt-handler processes can be compromised via the Service Account? A kind of cluster wide privilege escalation?

**Comment by xpivarc** [comment]
> Thanks @xpivarc. I am a bit confused about this line:
> 
> `If a malicious user has taken over a Kubernetes node where virt-handler (the KubeVirt node-daemon) is running, the virt-handler service account can be used to modify all node specs.`
> 
> Please could you clarify it? If a node is compromised then all virt-handler processes can be compromised via the Service Account? A kind of cluster wide privilege escalation?

You possibly could because you could obtain a Service account with `exec` rights.

**Comment by fabiand** [comment]
I agree with @alaypatel07  @xpivarc this change seems to be signifincant enough to have a deisng proposal in order to undertsand the motivation, and specifically tradeoffs with other implmenetations.

/hold

Until we have the design.

**Comment by EdDev** [comment]
> I agree with @alaypatel07 @xpivarc this change seems to be signifincant enough to have a deisng proposal in order to undertsand the motivation, and specifically tradeoffs with other implmenetations.
> 
> /hold
> 
> Until we have the design.

AFAIK, the PR already references a design proposal.
@xpivarc , it is at the moment marked as "draft". Do you think it is ready for review or you need to invest in it a bit more?

**Comment by xpivarc** [comment]
> I agree with @alaypatel07 @xpivarc this change seems to be signifincant enough to have a deisng proposal in order to undertsand the motivation, and specifically tradeoffs with other implmenetations.
> 
> /hold
> 
> Until we have the design.

Please, see the description.

>@xpivarc , it is at the moment marked as "draft". Do you think it is ready for review or you need to invest in it a bit more?

un-drafted, I am happy to work on it more if anybody finds it as not good enough.

File: cmd/virt-handler/virt-handler.go [hunk 12]
```diff
@@ -239,9 +239,22 @@ func (app *virtHandlerApp) Run() {
...
 
+	node, err := app.virtCli.CoreV1().Nodes().Get(context.TODO(), app.HostOverride, metav1.GetOptions{})
+	if err != nil {
+		panic(fmt.Sprintf("Failed to register node, %s", err))
+	}
+
 	_, err = app.virtCli.ShadowNodeClient().Create(context.TODO(), &v1.ShadowNode{
 		ObjectMeta: metav1.ObjectMeta{
 			Name: app.HostOverride,
+			OwnerReferences: []metav1.OwnerReference{
```
Comment (RamLavi) [inline cmd/virt-handler/virt-handler.go:250]: Hey @xpivarc - shouldn't the owner be the virt-handler pod itself?

**Comment by fabiand** [comment]
> Please, see the description.

Sorry, I was nt expressive enough: Let's hold until we have a merged design.

**Comment by xpivarc** [comment]
Replaced by https://github.com/kubevirt/kubevirt/pull/11982
